packages/grid/x-data-grid/src/components/GridRow.tsx

```diff
@@ -322,27 +335,38 @@ const GridRow = React.forwardRef<HTMLDivElement, GridRowProps>(function GridRow(
     }
   }
 
-  const style = {
-    ...styleProp,
-    maxHeight: rowHeight === 'auto' ? 'none' : rowHeight, // max-height doesn't support "auto"
-    minHeight,
-  };
+  const style = React.useMemo(() => {
+    const rowStyle = {
+      ...styleProp,
+      maxHeight: rowHeight === 'auto' ? 'none' : rowHeight, // max-height doesn't support "auto"
+      minHeight,
+    };
+    if (isNotVisible) {
+      return {
+        opacity: 0,
+        width: 0,
+        height: 0,
+      };
+    }
```

Comment:
Move the `if/return` to the start of this function. Apply the same pattern in `GridCell` where appropriate.